### PR TITLE
comparing Maps, find keys similar to missing (#27)

### PR DESCRIPTION
**Pull request by AlexCue987**
When keys are data classes with several fields, it really helps to find similarities between actual and expected keys, as shown in the section "Possible matches for missing keys" of the following output, where the matcher finds the most similar key:
```
val sweetGreenApple = Fruit("apple", "green", "sweet")
val sweetGreenPear = Fruit("pear", "green", "sweet")

            val e = shouldThrow<AssertionError> {
               mapOf(
                  sweetGreenApple to 1
               ) should containExactly(mapOf(sweetGreenPear to 1))
            }
            e.message shouldBe """
            |
            |Expected:
            |  mapOf(Fruit(name=apple, color=green, taste=sweet) to 1)
            |should be equal to:
            |  mapOf(Fruit(name=pear, color=green, taste=sweet) to 1)
            |but differs by:
            |  missing keys:
            |    Fruit(name=pear, color=green, taste=sweet)
            |  extra keys:
            |    Fruit(name=apple, color=green, taste=sweet)
            |
            |Possible matches for missing keys:
            |
            | expected: Fruit(name=pear, color=green, taste=sweet),
            |  but was: Fruit(name=apple, color=green, taste=sweet),
            |  distance: 0.67,
            |  fields:
            |  name expected: "pear", but was: "apple"
            |  color = "green"
            |  taste = "sweet"
              """.trimMargin()
```

### CODE REVIEW

**Comment by Kantis** [comment]
I think adding 700 lines of code to support slightly improved failure messages for _one assertion_ when working with just _one particular type_ is simply too much code doing too little. I say this because I feel there's already a lot of things to maintain in this project, and not enough maintainers.

However, I do think that it _is_ a nice feature for users, and your code is well structured. Do you have any plans for how the similarity checks can be used for more than map entries?

**Comment by AlexCue987** [comment]
> I think adding 700 lines of code to support slightly improved failure messages for _one assertion_ when working with just _one particular type_ is simply too much code doing too little. I say this because I feel there's already a lot of things to maintain in this project, and not enough maintainers.
> 
> However, I do think that it _is_ a nice feature for users, and your code is well structured. Do you have any plans for how the similarity checks can be used for more than map entries?

Yes, I'd like to use the same search for similarities to the following matchers:

* `shouldContain` 
* `shouldContainExactlyInAnyOrder`
* `shouldContainAllInAnyOrder`
* `shouldExistInOrder`
* `shouldContainAnyOf`
* `shouldBeIn`

Because we frequently match collections of data classes with multiple fields, and this additional information helps us faster grasp what exactly is different.

Thank you for quick response!

**Comment by AlexCue987** [comment]
There are multiple matchers where this search for similar elements would be very useful. And it won't be difficult to add them in time for the next release. But I'd like to add them later, in small separate PRs, if this one is approved/merged.
Does it sound like a good plan @Kantis ?

File: kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt [hunk 1]
```diff
@@ -269,6 +269,35 @@ class MapMatchersTest : WordSpec() {
...
+            |  missing keys:
+            |    Fruit(name=pear, color=green, taste=sweet)
+            |  extra keys:
+            |    Fruit(name=apple, color=green, taste=sweet)
+            |
+            |Possible matches for missing keys:
+            |
+            | expected: Fruit(name=pear, color=green, taste=sweet),
+            |  but was: Fruit(name=apple, color=green, taste=sweet),
+            |  similarity: 67.00%,
```
Comment (Kantis) [inline kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt]: I don't think users really care for this score. I suggest we just remove it to make output more concise

File: kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt [hunk 2]
```diff
@@ -269,6 +269,35 @@ class MapMatchersTest : WordSpec() {
...
+            |  fields:
+            |  name expected: "pear", but was: "apple"
+            |  color = "green"
+            |  taste = "sweet"
```
Comment (Kantis) [inline kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt]: Matching keys don't really matter, lets focus on the mismatch

File: kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt [hunk 3]
```diff
@@ -269,6 +269,35 @@ class MapMatchersTest : WordSpec() {
...
+            |  extra keys:
+            |    Fruit(name=apple, color=green, taste=sweet)
+            |
+            |Possible matches for missing keys:
+            |
+            | expected: Fruit(name=pear, color=green, taste=sweet),
+            |  but was: Fruit(name=apple, color=green, taste=sweet),
+            |  similarity: 67.00%,
+            |  fields:
+            |  name expected: "pear", but was: "apple"
```
Comment (Kantis) [inline kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt]:
```suggestion
            |  The following fields did not match:
            |     name: expected <"pear">, but was: <"apple">
```

Comment (Kantis) [reply kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt]: Increased indent for increased level of detail, and follow the `expected: <expected.print()>, but was <actual.print()>`

**Review by Kantis — CHANGES_REQUESTED** [review]
Alright, sounds good. Just have some feedback on the output format 🙏

File: kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt (hunk 3 above)
Comment (AlexCue987) [reply kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt]: > Increased indent for increased level of detail, and follow the `expected: <expected.print()>, but was <actual.print()>`

done

File: kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt (hunk 2 above)
Comment (AlexCue987) [reply kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt]: > Matching keys don't really matter, lets focus on the mismatch

done

File: kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt (hunk 1 above)
Comment (AlexCue987) [reply kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/maps/MapMatchersTest.kt]: > I don't think users really care for this score. I suggest we just remove it to make output more concise

done

**Review by Kantis — APPROVED** [review]

**Pull request merged**